vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h

```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_ENDPOINTS_REQUEST_FOR_H_
+#define BRAVE_VENDOR_BAT_NATIVE_LEDGER_SRC_BAT_LEDGER_INTERNAL_ENDPOINTS_REQUEST_FOR_H_
+
+#include <type_traits>
+#include <utility>
+
+#include "base/callback.h"
+#include "bat/ledger/internal/endpoints/request_builder.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/logging/logging.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace ledger::endpoints {
+
+template <typename Endpoint>
+class RequestFor {
+ public:
+  template <typename... Ts>
+  RequestFor(LedgerImpl* ledger, Ts&&... ts)
+      : ledger_(ledger),
+        request_(Endpoint{ledger, std::forward<Ts>(ts)...}.Request()) {
+    static_assert(std::is_base_of_v<RequestBuilder, Endpoint>,
+                  "Endpoint should be derived from RequestBuilder!");
+    DCHECK(ledger_);
+  }
+
+  RequestFor(const RequestFor&) = delete;
+  RequestFor& operator=(const RequestFor&) = delete;
+
+  RequestFor(RequestFor&&) = delete;
+  RequestFor& operator=(RequestFor&&) = delete;
+
+  void Send(base::OnceCallback<void(typename Endpoint::Result&&)> callback) && {
+    if (!request_ || !*request_) {
+      BLOG(0, "Failed to create request!");
+      return std::move(callback).Run(
```

Comment:
This should probably be run in a future turn, in order to avoid reentrancy surprises for callers. The basic pattern is:

```cpp
base::SequencedTaskRunnerHandle::Get()->PostTask(
    FROM_HERE, 
    base::BindOnce(std::move(callback), Endpoint::Error::FAILED_TO_CREATE_REQUEST));
```

Reply:
Hmm, would you mind elaborating on this a bit?

Reply:
Linking our [conversation](https://github.com/brave/brave-core/pull/15108#discussion_r972402047) here from your PR.